Objects/dictobject.c

```diff
@@ -6624,27 +6624,54 @@ make_dict_from_instance_attributes(PyInterpreterState *interp,
 }
 
 PyDictObject *
-_PyObject_MakeDictFromInstanceAttributes(PyObject *obj)
+_PyObject_MaterializeManagedDict(PyObject *obj)
 {
+    PyDictObject *dict = _PyObject_GetManagedDict(obj);
+    if (dict != NULL) {
+        return dict;
+    }
+
+    Py_BEGIN_CRITICAL_SECTION(obj);
+
+#ifdef Py_GIL_DISABLED
+    dict = _PyObject_GetManagedDict(obj);
+    if (dict != NULL) {
+        // We raced with another thread creating the dict
+        goto exit;
+    }
+#endif
+
+    OBJECT_STAT_INC(dict_materialized_on_request);
+
     PyDictValues *values = _PyObject_InlineValues(obj);
     PyInterpreterState *interp = _PyInterpreterState_GET();
     PyDictKeysObject *keys = CACHED_KEYS(Py_TYPE(obj));
     OBJECT_STAT_INC(dict_materialized_on_request);
-    return make_dict_from_instance_attributes(interp, keys, values);
+    dict = make_dict_from_instance_attributes(interp, keys, values);
+
+    FT_ATOMIC_STORE_PTR_RELEASE(_PyObject_ManagedDictPointer(obj)->dict,
+                                (PyDictObject *)dict);
+
+#ifdef Py_GIL_DISABLED
+exit:
+#endif
+    Py_END_CRITICAL_SECTION();
+    return dict;
 }
 
 
 
-int
-_PyObject_StoreInstanceAttribute(PyObject *obj, PyDictValues *values,
+static int
+store_instance_attr_lock_held(PyObject *obj, PyDictValues *values,
```

Comment:
I'll change it to do the insert here.  I was originally just keeping the caller of `_PyObject_TryStoreInstanceAttribute` the same where it's checking if the values are still valid, but there's no reason why we can't just handle the store to the dict.

I would love for it to be simpler, the double lock dance feels really ugly, but I'm not sure how we can handle it otherwise.  We could do something like get back to the object with a function which does the inverse of `_PyObject_InlineValues` to lock the object from the dict, but I think we'd still need to lock the dict to not race with detaching, so it doesn't help much (and might just complicate all of the mutations that can come from the dictionary APIs).  So I'll add a comment.